### DISCUSSION TITLE
Generalized transposed forms

### DISCUSSION THREAD

**Discussion by jpcima** [grame-cncm/faustlibraries#92] [Ideas]
What about transposed form II filters of arbitrary order?
This would be a variant of `fi.iir`, that generates transposed forms instead of direct forms.

![math_f9a](https://user-images.githubusercontent.com/17614485/130333934-a155503a-6325-4040-8ee4-27a6b8cb8d45.gif)

**Comment by sletz** [comment]
@josmithiii what do you think?

**Reply by josmithiii** [reply]
That sounds great to me.  Already we have the second-order case covered in filters.lib:

fi.tf21t = "fi.tf2 direct-form 1 transposed"

fi.tf22t = "fi.tf2 direct-form 2 transposed"

so, we could add fi.iir1t and fi.iir2t, and perhaps define fi.iirt = fi.iir2t, since that's the one you probably want (and what's drawn in the diagram above).

**Comment by sletz** [comment]
@jpcima any PR to suggest?

**Reply by jpcima** [reply]
I lack knowledge about how to implement this generically, but I'll take advice if there is.

**Comment by josmithiii** [comment]
The reason I haven't done iir2t myself is that I haven't needed it.  When I am under enough numerical stress to need transposed direct-form II, it's also been beneficial to factor into biquads, so that tf22t suffices.  Do you have a use case where iir2t would be best?

**Reply by jpcima** [reply]
At this stage, no I don't, it's only experimenting. In particular, the Elliptic and Chebyshev type II filters.
The page "Numerical Robustness of TDF-II" makes an explicit mention of the former.

**Reply by josmithiii** [reply]
As far as I know, everybody uses series biquads for such direct-form filter types.  Usually you first have to factor into biquads, and then you can implement those in TDFii which usually helps further.  However, one should also consider the allpass-decomposition forms.  The names here I think are Vaidyanathan, Regalia, and Mitra.  There's a great wrap-up here:

```
@BOOK{Vaidyanathan93,
        AUTHOR = "P. P. Vaidyanathan",
        TITLE = "Multirate Systems and Filter Banks",
        PUBLISHER = "Prentice-Hall",
        YEAR = 1993
}

```
